### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by MaxGraey**
This modification speed up method over 2.5x compare to original.
+ Iterate objects checking

### CODE REVIEW

**Comment by sindresorhus** [comment]
I appreciate the PR, but it's going to require some proof. 2.5x speedup is meaningless without context, you might have 2.5x speedup on a small array, but 50x slow down on a larger.

https://jsperf.com

File: index.js [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
-		const a = arr[i];
-		const b = arr[i + 1];
 
-		if (comp ? comp(a, b) > 0 : a > b) {
+const defaultComparator = (a, b) => a > b ? 1 : 0;
+module.exports = (arr, comp) => {
+	// pass only iterable objects
+	if (!arr || typeof arr[Symbol.iterator] !== 'function') {
+		return false;
+	}
```
Comment (sindresorhus) [inline index.js:8]: This is unrelated to this change.

File: index.js [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
+	comp = comp || defaultComparator;
+	var a = arr[0];
+	for (var i = 1, len = arr.length; i < len; i++) {
+		var b = arr[i];
+		if (comp(a, b) > 0) {
```
Comment (sindresorhus) [inline index.js:13]: I have a gut feeling the previous code here was faster as it only called `comp` when actually specified.

File: index.js [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
-		if (comp ? comp(a, b) > 0 : a > b) {
+const defaultComparator = (a, b) => a > b ? 1 : 0;
+module.exports = (arr, comp) => {
+	// pass only iterable objects
+	if (!arr || typeof arr[Symbol.iterator] !== 'function') {
+		return false;
+	}
+	comp = comp || defaultComparator;
+	var a = arr[0];
+	for (var i = 1, len = arr.length; i < len; i++) {
```
Comment (sindresorhus) [inline index.js:11]: Caching length hasn't been necessary for years.

File: index.js [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
 
-		if (comp ? comp(a, b) > 0 : a > b) {
+const defaultComparator = (a, b) => a > b ? 1 : 0;
+module.exports = (arr, comp) => {
+	// pass only iterable objects
+	if (!arr || typeof arr[Symbol.iterator] !== 'function') {
+		return false;
+	}
+	comp = comp || defaultComparator;
+	var a = arr[0];
```
Comment (sindresorhus) [inline index.js:10]: I'm using let/const, not interested in using `var`, even for performance reasons. V8 will optimize those soon enough.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (MaxGraey) [reply index.js:8]: I agree, but this is an important condition, otherwise if the input comes non-iterable or empty object, we get a failure

File: index.js (hunk 4 above)
Comment (MaxGraey) [reply index.js:10]: But code written exclusively for Node 4/6 as I understand it. It therefore rejected [this PR ](https://github.com/sindresorhus/is-array-sorted/pull/1)

File: index.js (hunk 1 above)
Comment (sindresorhus) [reply index.js:8]: Yes, that's the idea. You shouldn't just silently return false when you get the wrong type.

File: index.js (hunk 4 above)
Comment (sindresorhus) [reply index.js:10]: Node.js 4, which this module targets, supports let/const fine, but not default parameters.

http://node.green/ is a good resource.

File: index.js (hunk 3 above)
Comment (MaxGraey) [reply index.js:11]: Actually it [still has](https://jsperf.com/array-length-vs-cached/54). 15% boost in my Chome 55 (V8 5.5)

Comment (sindresorhus) [reply index.js:11]: And in this it's the same: https://jsperf.com/caching-array-length/4

It's only microbenchmarks anyways. Not very realistic.

http://mrale.ph/blog/2014/12/24/array-length-caching.html

**Comment by MaxGraey** [comment]
Just see this (https://jsperf.com/is-array-sorted/1)

screenshot

File: index.js (hunk 3 above)
Comment (MaxGraey) [reply index.js:11]: I created fork of this test because original has very small length. So we still have 5-6% improvements https://jsperf.com/caching-length-2/1

File: index.js (hunk 4 above)
Comment (MaxGraey) [reply index.js:10]: Yes but on Node.js 4 and even 7 we have very unoptimized 'let' statement

File: index.js (hunk 3 above)
Comment (sindresorhus) [reply index.js:11]: 5% gain on a microbench is honestly not worth it.

Comment (MaxGraey) [reply index.js:11]: But to agree from this micro optimization of the code does not become longer or non-clear. So why would not implement this? And what you think about this test (https://jsperf.com/is-array-sorted/1)?

Comment (sindresorhus) [reply index.js:11]: Because it's just meaningless churn. I could also write it in C, compile it as a native Node.js addon, and use that, but it wouldn't make a difference in actual usage. Micro-benchmarks do not reflect actual usage.

**Comment by sindresorhus** [comment]
I may be ok with the following if you can show it's considerably faster than master:

```js
'use strict';
module.exports = (arr, comp) => {
	let a = arr[0];

	for (let i = 1; i < arr.length; i++) {
		let b = arr[i];

		if (comp ? comp(a, b) > 0 : a > b) {
			return false;
		}

		a = b;
	}

	return true;
};
```

**Comment by MaxGraey** [comment]
@sindresorhus I'm a little puzzled. I have created benchmark tests on [jsPerf](https://jsperf.com/is-array-sorted/1)(See setup code) which showed a significant performance even on huge arrays in comparison with the old method. How can I even prove it? =)

PS If you look at the code, you see the obvious things:
- Each iteration there is only one reading from the array instead of two
- Simplified the logic of comparison
#### Bonus
- Validation of input array

**Comment by sindresorhus** [comment]
> I have created benchmark tests on jsPerf(See setup code) which showed a significant performance even on huge arrays in comparison with the old method. How can I even prove it?

Change the JSfiddle to use the above code.

> Each iteration there is only one reading from the array instead of two

That change I'm ok with if it proves to be faster.

> Simplified the logic of comparison

Already answered: https://github.com/sindresorhus/is-array-sorted/pull/2#discussion_r97787658

> Bonus: Validation of input array

It's not relevant to this PR.

File: index.js (hunk 3 above)
Comment (MaxGraey) [reply index.js:11]: Nope, native c++ code will be slower for this case because we would get bottleneck on bridging between C++ and JS. I know it because wrote native module for V8 for deep cloning. That's why some build-in function such as ```Array.proptotype.sort``` (https://github.com/v8/v8/blob/950d2051a5ff065a5bc1d31f0e5d1bba850d0b3c/src/array.js) wrote on JS-side into private area

Comment (sindresorhus) [reply index.js:11]: Sorry, my comment wasn't meant to be taken literally.

Comment (MaxGraey) [reply index.js:11]: I understand this. I closed the PR because I see no more sense to prove the relevance of each line
